dbms/src/Disks/DiskS3.h

```diff
@@ -0,0 +1,141 @@
+#pragma once
+
+#include <Common/config.h>
+
+#if USE_AWS_S3
+#    include "DiskFactory.h"
+
+#    include <aws/s3/S3Client.h>
+#    include <Poco/DirectoryIterator.h>
+
+
+namespace DB
+{
+class DiskS3 : public IDisk
+{
+public:
+    friend class DiskS3Reservation;
+
+    DiskS3(String name_, std::shared_ptr<Aws::S3::S3Client> client_, String bucket_, String s3_root_path_, String metadata_path_);
+
+    const String & getName() const override { return name; }
+
+    const String & getPath() const override { return s3_root_path; }
+
+    ReservationPtr reserve(UInt64 bytes) override;
+
+    UInt64 getTotalSpace() const override { return std::numeric_limits<UInt64>::max(); }
+
+    UInt64 getAvailableSpace() const override { return std::numeric_limits<UInt64>::max(); }
+
+    UInt64 getUnreservedSpace() const override { return std::numeric_limits<UInt64>::max(); }
+
+    UInt64 getKeepingFreeSpace() const override { return 0; }
+
+    bool exists(const String & path) const override;
+
+    bool isFile(const String & path) const override;
+
+    bool isDirectory(const String & path) const override;
+
+    size_t getFileSize(const String & path) const override;
+
+    void createDirectory(const String & path) override;
+
+    void createDirectories(const String & path) override;
+
+    void clearDirectory(const String & path) override
+    {
+        for (auto it{iterateDirectory(path)}; it->isValid(); it->next())
+            remove(it->path());
+    }
```

Comment:
The `clearDirectory` method should removes only files (see https://github.com/ClickHouse/ClickHouse/pull/8356#discussion_r362129176). Probably we need to add some option to `clearDirectory` for removing directories as well.